cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/tracing/GcpTracingInterceptor.java

```diff
@@ -0,0 +1,124 @@
+package com.sequenceiq.cloudbreak.cloud.gcp.tracing;
+
+import java.util.Map;
+
+import javax.inject.Inject;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.http.HttpHost;
+import org.apache.http.HttpRequest;
+import org.apache.http.HttpRequestInterceptor;
+import org.apache.http.HttpResponse;
+import org.apache.http.HttpResponseInterceptor;
+import org.apache.http.RequestLine;
+import org.apache.http.StatusLine;
+import org.apache.http.protocol.HttpContext;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.tracing.TracingUtil;
+import com.sequenceiq.cloudbreak.util.Benchmark;
+
+import io.opentracing.References;
+import io.opentracing.Scope;
+import io.opentracing.Span;
+import io.opentracing.Tracer;
+
+@Component
+public class GcpTracingInterceptor implements HttpRequestInterceptor, HttpResponseInterceptor {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GcpTracingInterceptor.class);
+
+    private static final String JAVA_GCP_SDK = "java-gcp-sdk";
+
+    private static final int MIN_OK_STATUS_CODE = 200;
+
+    private static final int MAX_OK_STATUS_CODE = 300;
+
+    private static final ThreadLocal<Pair<Scope, Span>> DATA = new ThreadLocal<>();
+
+    @Inject
+    private Tracer tracer;
+
+    @Override
+    public void process(HttpRequest request, HttpContext context) {
+        StackWalker stackWalker = StackWalker.getInstance();
+        StringBuilder apiOperation = new StringBuilder();
+        Benchmark.measure(() -> getApiOperationFromStackTrace(stackWalker, apiOperation),
+                LOGGER, "STACK WALKING TOOK: {}ms");
+        activateNewTracingSpan(request, context, apiOperation);
+    }
+
+    private void activateNewTracingSpan(HttpRequest request, HttpContext context, StringBuilder apiOperation) {
+        RequestLine requestLine = request.getRequestLine();
+        String operationName = createOperationName(context, apiOperation, requestLine);
+        Span span = initSpan(requestLine, operationName);
+        Scope scope = tracer.activateSpan(span);
+        DATA.set(Pair.of(scope, span));
+    }
+
+    private void getApiOperationFromStackTrace(StackWalker stackWalker, StringBuilder apiOperation) {
+        stackWalker.forEach(stackFrame -> {
```

Comment:
To optimize please rewrite as a `for` statement and exit loop after the isFirstGcpApiMethod was true.
This way you do not need to unnecessarily walk through the whole call stack.